File: common/python/quartz/data/table.py
from quartz.data.base import Database
from quartz.log import Log
from quartz.error import Error
from abc import ABC, abstractmethod


class Table(ABC):

    class Info:
        def __init__(self, table:str, columns:list=None, n_keys:int=None, replace:bool=False, tag:str=None):
            self.table = table
            self.replace = replace
            self.tag = tag
            if columns is None: columns = []
            if (n_keys is None) or (n_keys >= len(columns)):
                self.keys = columns
                self.values = []
            else:
                self.keys = columns[:n_keys]
                self.values = columns[n_keys:]
            # c = ', '.join([ "∅" if c is None else str(c) for c in columns])
            # k = ', '.join([ "∅" if c is None else str(c) for c in self.keys])
            # v = ', '.join([ "∅" if c is None else str(c) for c in self.values])
            # Log.debug(f"[{c}] =({(n_keys is None) and '*' or str(n_keys)})=> [{k}] [{v}]")

        def __str__(self):
            k = ', '.join([ "∅" if c is None else str(c) for c in self.keys])
            v = ', '.join([ "∅" if c is None else str(c) for c in self.values])
            r = ' R' if self.replace else ''
            t = f'<{self.tag}>' if (self.tag is not None) else ''
            return f'"{self.table}"{t}[{k}][{v}]'
        
    def __init__(self, name:str, db:Database):
        if not isinstance(name, str): Error.invalid("table name")
        if not isinstance(db, Database): Error.invalid("table database")
        self._name = name
        self._db = db        

    @property
    def name(self):
        return self._name

    @property
    def db(self):
        return self._db

    def insert(self, x, debug=False):
        info = x if isinstance(x, Table.Info) else self._insert(x)
        if not isinstance(info, Table.Info):
            Error.invalid(f'"{self.name}" insert', x)
        if info.tag is None:
            # INSERT INTO book(title, year) VALUES ('Dune', 1965)
            # REPLACE INTO book(title, year) VALUES ('Dune', 1965)
            action = info.replace and "REPLACE" or "INSERT"
            columns = info.keys + info.values
            names = ", ".join([ f"`{c[0]}`" for c in columns ])
            args = [ c[1] for c in columns ]
            marks = ", ".join([ "?" ] * len(columns) )
            query = f"{action} {info.table}({names}) VALUES({marks})"
            return self.db.exec(query, args, debug)
        else:
            # @book/insert|insert_id|replace
            args = info.values if ("insert" == info.tag) else (info.keys + info.values)
            query = f"@{info.table}/{info.tag}"
            return self.db.exec(query, args, debug)

    def update(self, x, debug=False):
        info = x if isinstance(x, Table.Info) else self._update(x)
        if not isinstance(info, Table.Info):
            Error.invalid(f'"{self.name}" update', x)
        if info.tag is None:
            # UPDATE book SET year=2000 WHERE id=1
            sets = ", ".join([ f"`{c[0]}`=?" for c in info.values ])
            where = " AND ".join([ f"`{k[0]}`=?" for k in info.keys ])
            query = f"UPDATE `{info.table}` SET {sets} WHERE {where}"
            args = [ c[1] for c in info.values ] + [ c[1] for c in info.keys ]
            self.db.exec(query, args, debug)
        else:
            # @book/update|replace
            query = f"@{info.table}/{info.tag}"
            args = info.values + info.keys
            self.db.exec(query, args, debug)           
        return x

    def delete(self, x, debug=False):
        info = x if isinstance(x, Table.Info) else self._delete(x)
        if not isinstance(info, Table.Info): 
            Error.invalid(f'"{self.name}" delete', x)
        if info.tag is None:
            # DELETE FROM book WHERE year >= ? AND year <= ?
            query = f"DELETE FROM `{info.table}`"
            where = []
            args = []
            for i in range(len(info.keys)):
                if info.keys[i][1] is not None:
                    where.append(f"{info.keys[i][0]}=?")
                    args.append(info.keys[i][1])
            if len(args) > 0:
                query += " WHERE " + " AND ".join(where)
                self.db.exec(query, args, debug)
        else:
            query = f"@{info.table}/delete-{info.tag}"
            self.db.exec(query, info.keys, debug)
        return x

    def deleteAll(self, debug=False):
        self.db.exec(f"@{self.table}/delete", [], debug)

    def begin(self, x=None, columns:str='full', order:list[str]=None, limit:int=None, debug:bool=False):
        [query, args] = self.preselect(x, columns, order, limit)
        return self.db.begin(query, args, debug=debug)

    def next(self, cursor, builder=None):
        row = self.db.next(cursor)
        if ((isinstance(row, tuple) or isinstance(row, list)) and (len(row) > 0)):
            return builder(row) if builder else row
        return None
    
    def one(self, x=None, builder=None, columns:str='full', order:list[str]=None, limit:int=None, debug:bool=False):
        cur = self.begin(x, columns, order, limit, debug)
        return self.next(cur, builder)

    def select(self, x=None, builder=None, columns:str='full', order:list[str]=None, limit:int=None, debug:bool=False) -> list:
        [query, args] = self.preselect(x, columns, order, limit)
        rows = self.db.all(query, args, debug=debug)
        if (isinstance(rows, list) and (len(rows) > 0)):
            return [ builder(r) for r in rows ] if builder else rows
        return []

    def preselect(self, x=None, columns:list[str]='full', order:str=None, limit:str=None) -> list:
        info = x if isinstance(x, Table.Info) else self._select(x)
        if not isinstance(info, Table.Info): 
            Error.invalid(f'"{self.name}" select', x)
        ord = "" if (order is None) else f" ORDER BY {order}"
        lim = "" if (limit is None) else f" LIMIT {lim}"
        tail = ord + lim
        if info.tag is None:
            # SELECT id, title FROM book WHERE author = ? AND year > ?
            nonkeys = ", ".join([ f"`{c[0]}`" for c in info.values ])
            where = " AND ".join([ f"`{c[0]}`=?" for c in info.keys ])
            args = [ c[1] for c in info.keys ]
            query = f"SELECT {nonkeys} FROM `{info.table}` WHERE {where}{tail}"
            return [ query, args ]
        else:
            # @book/select_<tag>
            tag = 'all' if (info.tag is None) else info.tag
            query = self.db.sql(f"@{info.table}/select_{columns}-{tag}") + tail
            return [ query, info.keys ]

    @abstractmethod
    def _insert(self, x) -> Table.Insert:
        return None

    @abstractmethod
    def _update(self, x) -> Table.Update:
        return None

    @abstractmethod
    def _delete(self, x) -> Table.Delete:
        return None

    @abstractmethod
    def _select(self, x) -> Table.Select:
        return None

File: common/php/data/table.php
<?php
namespace quartz;
require_once __DIR__ . '/base.php';
require_once __DIR__ . '/query.php';

class Info {
    public $table;
    public $values;
    public $replace;

    function __construct(string $table, ?array $values=null, bool $replace=false) {
        $this->table = $table;
        $this->values = is_null($values)? [] : $values;
        $this->replace = $replace;
    }
}

class TagInfo extends Info {
    public $tag;

    function __construct(string $table, string $tag, ?array $values=null, bool $replace=false) {
        parent::__construct($table, $values, $replace);
        $this->tag = $tag;
    }
}

class KeyInfo extends Info {
    public $keys;

    function __construct(string $table, ?array $keys=null, ?array $values=null, bool $replace=false) {
        parent::__construct($table, $values, $replace);
        $this->keys = $keys;
    }
}


abstract class Table {
    public string $name;
    public Database $db;

    function __construct(string $name, Database $db) {
        $this->name = $name;
        $this->db = $db;
    }

    function insert($x, bool $debug=false) {
        $info = ($x instanceof Info) ? $x : $this->_insert($x);
        if ($info instanceof KeyInfo) {
            # INSERT INTO book(title, year) VALUES ('Dune', 1965)
            # REPLACE INTO book(title, year) VALUES ('Dune', 1965)
            $action = $info->replace ? "REPLACE" : "INSERT";
            $columns = array_merge($info->keys, $info->values);
            $names = implode(', ', array_map(fn($c) => "`{$c[0]}`", $columns));
            $args = array_map(fn($c) => $c[1], $columns);
            $marks = array_fill(0, count($columns), '?');
            $query = "{$action} $info->table($names) VALUES($marks)";
            return $this->db->exec($query, $values, $debug);
        }
        elseif ($info instanceof Info) {
            # @book/insert|insert_id|replace
            $args = $info->values;
            if ($info->replace) {
                $query = "@{$info->table}/replace";
            }
            elseif (is_null($info->values[0])) {
                $query = "@{$info->table}/insert";
                array_shift($args);
            }
            else {
                $query = "@{$info->table}/insert_id";
            }
            return $this->db->exec($query, $args, $debug);
        }
        else Error::invalid("insert \"{$this->name}\"", $x);
    }

    function update($x, bool $debug=false) {
        $info = ($x instanceof Info) ? $x : $this->_update($x);
        if ($info instanceof KeyInfo) {
            # UPDATE book SET year=2000 WHERE id=1
            $sets = implode(', ', array_map(fn($c) => "`{$c[0]}`", $info->values));
            $where = implode(' AND ', array_map(fn($c) => "`{$c[0]}`=?", $info->keys));
            $query = "UPDATE `{$info->table}` SET $sets WHERE $where";
            $args = array_merge(array_map(fn($c) => $c[1], $info->values), array_map(fn($c) => $c[1], $info->keys));
            $this->db->exec($query, $args, $debug);
        }
        elseif ($info instanceof Info) {
            # @book/update|replace
            $query = "@{$info->table}/" . ($info->replace ? "replace" : "update");
            $this->db->exec($query, $info->values, $debug);
        }
        else Error::invalid("update \"{$this->name}\"", $x);
        return $x;
    }

    function delete($x, bool $debug=false) {
        $info = ($x instanceof Info) ? $x : $this->_delete($x);
        if ($info instanceof KeyInfo) {
            # DELETE FROM book WHERE year >= ? AND year <= ?
            $query = "DELETE FROM `{$info->table}`";
            $where = [];
            $args = [];
            foreach ($info->keys as $i => $key) {
                if (!is_null($info->keys[$i][1])) {
                    $where[] = "{$info->keys[i][0]}=?";
                    $args[] = $info->keys[i][1];
                }
            }
            if (count($args) > 0) {
                $query .= " WHERE " . implode(' AND ', $where);
            }
            $this->db->exec($query, $args, $debug);
        }
        elseif ($info instanceof TagInfo) {
            $query = "@{$info->table}/delete_{$info->tag}";
            $this->db->exec($query, $info->values, debug);
        }
        else Error::invalid("delete \"{$this->name}\"", $x);
        return $x;
    }

    function deleteAll($debug=false) {
        $this->db->exec("@{$this->table}/delete", [], $debug);
    }

    function begin($x=null, ?string $columns='full', ?array $order=null, ?string $limit=null, bool $debug=false) {
        [$query, $args] = $this->preselect($x, $columns, $order, $limit);
        return $this->db->begin($query, $args, debug:$debug);
    }

    function next($cursor, ?callable $builder=null) {
        $row = $this->db->next($cursor);
        if (is_array($row) && (count($row) > 0)) {
            return is_null($builder) ? $row : $builder($row);
        }
        return null;
    }

    function one($x=null, ?callable $builder=null, ?string $columns='full', ?array $order=null, ?string $limit=null, bool $debug=false) {
        $cur = $this->begin($x, $columns, $order, $limit, $debug);
        return $this->next($cur, $builder);
    }

    function select($x=null, ?callable $builder=null, ?string $columns='full', ?array $order=null, ?string $limit=null, bool $debug=false): array {
        [$query, $args] = $this->preselect($x, $columns, $order, $limit);
        $rows = $this->db->all($query, $args, debug:$debug);
        if (is_array($rows) && (count($rows) > 0)) {
            return is_null($builder) ? $rows :  array_map(fn($r) => $builder($r), $rows);
        }
        return [];
    }

    private function preselect($x=null, ?string $columns='full', ?array $order=null, ?string $limit=null): array {
        $info = ($x instanceof Info) ? $x : $this->_select($x);

        $ord = is_null($order) ? "" : " ORDER BY " . implode(', ', array_map(fn($c) => "`$c`", $order));
        $lim = is_null($limit) ? "" : " LIMIT $lim";
        $tail = $ord . $limit;

        if ($info instanceof KeyInfo) {
            # SELECT id, title FROM book WHERE author = ? AND year > ?
            $nonkeys = implode(', ', array_map(fn($c) => "`{$c[0]}`", $info->values));
            $where = implode(' AND ', array_map(fn($c) => "`{$c[0]}`", $info->keys));
            $args = array_map(fn($c) => $c[1], $info->keys);
            $query = "SELECT $nonkeys FROM `{$info->table}` WHERE {$where}{$tail}";
            return [ $query, $args ];
        }
        elseif ($info instanceof TagInfo) {
            // @book/select_<tag>
            $tag = $info->tag ? $info->tag : 'all';
            $query = $this->db->sql("@{$info->table}/select_$tag-$columns") . $tail;
            return [ $query, $info->values ];
        }
        else Error::invalid("select \"{$this->name}\"", $x);
    }

    abstract public function _insert($x): ?Info;
    abstract public function _update($x): ?Info;
    abstract public function _delete($x): ?Info;
    abstract public function _select($x): ?Info;

}

?>